src/Platform/Microsoft.Testing.Platform/IPC/Serializers/CommandLineOptionMessagesSerializer.cs

```diff
@@ -0,0 +1,157 @@
+    {
+        RoslynDebug.Assert(stream.CanSeek, "We expect a seekable stream.");
+
+        CommandLineOptionMessages commandLineOptionMessages = (CommandLineOptionMessages)objectToSerialize;
+
+        WriteShort(stream, (short)GetFieldCount(commandLineOptionMessages));
+
+        WriteField(stream, CommandLineOptionMessagesFields.ModuleName, commandLineOptionMessages.ModuleName);
+        WriteCommandLineOptionMessagesPayload(stream, CommandLineOptionMessagesFields.CommandLineOptionMessageList, commandLineOptionMessages.CommandLineOptionMessageList);
+    }
+
+    private static void WriteCommandLineOptionMessagesPayload(Stream stream, short id, CommandLineOptionMessage[] commandLineOptionMessageList)
+    {
+        if (commandLineOptionMessageList is null)
+        {
+            return;
+        }
+
+        WriteShort(stream, id);
+
+        // We will reserve an int (4 bytes)
+        // so that we fill the size later, once we write the payload
+        WriteInt(stream, 0);
+
+        long before = stream.Position;
+        WriteInt(stream, commandLineOptionMessageList.Length);
+        foreach (CommandLineOptionMessage commandLineOptionMessage in commandLineOptionMessageList)
+        {
+            WriteShort(stream, (short)GetFieldCount(commandLineOptionMessage));
+
+            WriteField(stream, CommandLineOptionMessageFields.Name, commandLineOptionMessage.Name);
+            WriteField(stream, CommandLineOptionMessageFields.Description, commandLineOptionMessage.Description);
+            WriteField(stream, CommandLineOptionMessageFields.Arity, commandLineOptionMessage.Arity);
+            WriteField(stream, CommandLineOptionMessageFields.IsHidden, commandLineOptionMessage.IsHidden);
+            WriteField(stream, CommandLineOptionMessageFields.IsBuiltIn, commandLineOptionMessage.IsBuiltIn);
+        }
+
+        // NOTE: We are able to seek only if we are using a MemoryStream
+        // thus, the seek operation is fast as we are only changing the value of a property
+        WriteAtPosition(stream, (int)(stream.Position - before), before - sizeof(int));
+    }
+
+    private static int GetFieldCount(CommandLineOptionMessages commandLineOptionMessages)
+    {
+        return (IsNull(commandLineOptionMessages.ModuleName) ? 1 : 0) +
+           (commandLineOptionMessages.CommandLineOptionMessageList is null ? 0 : 1);
+    }
+
+    private static int GetFieldCount(CommandLineOptionMessage commandLineOptionMessage)
+    {
+        return (IsNull(commandLineOptionMessage.Name) ? 1 : 0) +
+            (IsNull(commandLineOptionMessage.Description) ? 1 : 0) +
+            (IsNull(commandLineOptionMessage.Arity) ? 1 : 0) +
+            (IsNull(commandLineOptionMessage.IsHidden) ? 1 : 0) +
+            (IsNull(commandLineOptionMessage.IsBuiltIn) ? 1 : 0);
+    }
+
+    private static bool IsNull<T>(T field)
+    {
+        return typeof(T) == typeof(string) ? field is not null : true;
```

Comment:
think that this check can be removed and used directly inside the GetFieldCount.

 A reference type can be null, like a string, a value type cannot, like boolean if it's a nullable Null<T> it's a ref type anyway and the is null check will work correctly.

```
     private static int GetFieldCount(CommandLineOptionMessage commandLineOptionMessage)
    {
        return (commandLineOptionMessage.Name is not null ? 1 : 0) +
        ...            
    }
```    
this will also take advantage of the compiler that will help to fail if we're comparing a value type to null, a value type cannot be null so the check is not needed at all, better parf.
(Also the name IsNull looks like you wanted to return the opposite so if it's not null.)